gateway/ln-gateway/src/lib.rs

```diff
@@ -1545,6 +1560,234 @@ pub(crate) async fn fetch_lightning_node_info(
+            payment_fees: self.payment_fees(),
+            outgoing_cltv_delta: OUTGOING_CLTV_DELTA_NG,
+        })
+    }
+
+    pub fn payment_fees(&self) -> PaymentFees {
+        PaymentFees {
+            // we take a fee of one percent for outgoing contracts
+            send: PaymentFee::default(),
+            // we take a fee of one percent for incoming contracts
+            receive: PaymentFee::default(),
+        }
+    }
+
+    async fn send_payment(
+        &self,
+        payload: SendPaymentPayload,
+    ) -> std::result::Result<std::result::Result<[u8; 32], Signature>, SendPaymentError> {
+        let clients = self.clients.read().await;
+
+        let client = clients
+            .get(&payload.federation_id)
+            .ok_or(SendPaymentError::UnknownFederationId)?
+            .value();
+
+        let operation_id = OperationId(payload.contract.contract_id().0.into_inner());
+
+        let module = client.get_first_module::<GatewayClientModuleV2>();
+
+        if client.operation_exists(operation_id).await {
+            return Ok(module.subscribe_send(operation_id, payload.contract).await);
+        }
+
+        if payload.contract.claim_pk != module.keypair.public_key() {
+            return Err(SendPaymentError::NotOurKey);
+        }
+
+        if payload.invoice.consensus_hash::<sha256::Hash>() != payload.contract.invoice_hash {
+            return Err(SendPaymentError::InvalidInvoiceHash);
+        }
+
+        let max_delay = module
+            .module_api
+            .outgoing_contract_expiration(&payload.contract.contract_id())
+            .await
+            .map_err(|_| SendPaymentError::FederationUnreachable)?
+            .ok_or(SendPaymentError::UnconfirmedContract)?
+            .saturating_sub(OUTGOING_CLTV_DELTA_NG);
+
+        if max_delay == 0 {
+            return Err(SendPaymentError::TimeoutTooClose);
+        }
+
+        let invoice_msats = payload
+            .invoice
+            .amount_milli_satoshis()
+            .ok_or(SendPaymentError::InvoiceMissingAmount)?;
+
+        let contract_amount = self.payment_fees().outgoing_contract_amount(invoice_msats);
+
+        if contract_amount != payload.contract.amount {
+            return Err(SendPaymentError::Unbalanced);
+        }
+
+        let max_fee_msat = (contract_amount.msats - invoice_msats) / 2;
+
+        module
+            .start_send_state_machine(
+                operation_id,
+                max_delay,
+                max_fee_msat,
+                payload.invoice,
+                payload.contract.clone(),
+            )
+            .await
+            .ok();
+
+        Ok(module.subscribe_send(operation_id, payload.contract).await)
+    }
+
+    async fn create_invoice(
+        &self,
+        payload: CreateInvoicePayload,
+    ) -> std::result::Result<Bolt11Invoice, CreateInvoiceError> {
+        if !payload.contract.verify() {
+            return Err(CreateInvoiceError::InvalidContract);
+        }
+
+        loop {
```

Comment:
We can use `autocommit` here instead of manually looping

Reply:
Also, instead of looping, you can just return `HashAlreadyRegistered` if `commit_tx_result` returns an error. On the second loop its going to return that anyway, it will be nice to get rid of the loop here.